src/models/history/tree-monitor.ts

```diff
@@ -1,163 +1,261 @@
 import { addDisposer, addMiddleware, 
-    getPath, 
-    getRunningActionContext, 
-    IActionContext, 
-    IJsonPatch, Instance, IPatchRecorder, isActionContextThisOrChildOf, 
-    isAlive, recordPatches } from "mobx-state-tree";
+  getPath, 
+  getRoot,
+  getRunningActionContext, 
+  IActionContext, 
+  IJsonPatch, Instance, IPatchRecorder, isActionContextThisOrChildOf, 
+  isAlive, recordPatches } from "mobx-state-tree";
 import { nanoid } from "nanoid";
 import { TreeManagerAPI } from "./tree-manager-api";
 import { TreePatchRecordSnapshot } from "./history";
 import { Tree } from "./tree";
 import { createActionTrackingMiddleware3, IActionTrackingMiddleware3Call } from "./create-action-tracking-middleware-3";
+import { DEBUG_UNDO } from "../../lib/debug";
 
 interface CallEnv {
-    recorder: IPatchRecorder;
-    sharedModelModifications: SharedModelModifications;
-    historyEntryId: string;
-    exchangeId: string;
-    undoable: boolean;
+  recorder: IPatchRecorder;
+  sharedModelModifications: SharedModelModifications;
+  historyEntryId: string;
+  exchangeId: string;
+  undoable: boolean;
 }
 
 type SharedModelModifications = Record<string, number>;
 
 const runningCalls = new WeakMap<IActionContext, IActionTrackingMiddleware3Call<CallEnv>>();
 
-export const addTreeMonitor = (tree: Instance<typeof Tree>,  manager: TreeManagerAPI, includeHooks: boolean) => {
-    const treeMonitorMiddleware = createActionTrackingMiddleware3<CallEnv>({
-        filter(call) {
-            if (call.env) {
-                // already recording
-                return false;
-            }
-            return true;
-        },
-        onStart(call) {
-            // Add this call to our global WeakMap so other functions can 
-            // modify these calls.
-            runningCalls.set(call.actionCall, call);
+function getActionName(call: IActionTrackingMiddleware3Call<CallEnv>) {
```

Comment:
It seems a bit odd that a function called `getActionName()` returns something more than the action name. Perhaps something like `getActionNameWithPath()` or `getFullActionName()`?

Reply:
Yeah I think actionPath is better